src/forklift/core.py

```diff
@@ -233,70 +190,99 @@ def _hash(crate, hash_path):
         #: truncate destination table since we are hashing for the first time
         arcpy.TruncateTable_management(crate.destination)
 
-    shape_token = 'SHAPE@'
-    is_table = _is_table(crate)
+    shape_token = 'SHAPE@WKT'
+    src_id_field = 'src_id'
 
     log.info('checking for changes...')
     #: finding and filtering common fields between source and destination
     fields = set([fld.name for fld in arcpy.ListFields(crate.destination)]) & set([fld.name for fld in arcpy.ListFields(crate.source)])
-    fields = _filter_fields(fields)
-
-    sql_clause = None
+    fields = _filter_fields(fields, crate)
 
     #: keep track of OID token in order to remove from hashing
-    att_hash_sub_index = None
-    if 'OID@' in fields:
-        att_hash_sub_index = -1
-        source_describe = arcpy.Describe(crate.source)
-        sql_clause = (None, 'ORDER BY {}'.format(source_describe.OIDFieldName))
+    primary_key_index = -1
+    sql_clause = (None, 'ORDER BY {}'.format(crate.source_primary_key))
 
-    if not is_table:
+    if not crate.is_table():
         fields.append(shape_token)
-        att_hash_sub_index = -2
+        primary_key_index = -2
+
+    changes = Changes(list(fields))
+    changes.table = crate.source
+    changes.fields.append('OID@')
 
-    changes = Changes(fields)
-    #: TODO update to use with pickle or geodatabase
     attribute_hashes, geometry_hashes = _get_hash_lookups(crate.name, hash_gdb_path)
     unique_salt = 0
 
+    insert_cursor = None
+    if needs_reproject:
+        changes.table = temp_table = arcpy.CreateFeatureclass_management(arcpy.env.scratchGDB,
+                                                                         crate.name,
+                                                                         geometry_type=crate.source_describe.shapeType.upper(),
+                                                                         template=crate.source,
+                                                                         spatial_reference=crate.source_describe.spatialReference)[0]
+        arcpy.AddField_management(temp_table, src_id_field, 'LONG')
```

Comment:
we might need to convert this to a string if a crates primary key was not a long. like a guid for example.